tapgarden/re-org_watcher.go

```diff
@@ -0,0 +1,424 @@
+package tapgarden
+
+import (
+	"context"
+	"fmt"
+	"sync"
+
+	"github.com/btcsuite/btcd/chaincfg/chainhash"
+	"github.com/lightninglabs/taproot-assets/fn"
+	"github.com/lightninglabs/taproot-assets/proof"
+	"github.com/lightningnetwork/lnd/chainntnfs"
+)
+
+// proofRegistration is a struct that holds all proofs that need to be watched
+// for re-orgs and a callback that will be called if a re-org happened.
+type proofRegistration struct {
+	proofs []*proof.Proof
+
+	updateCb ProofUpdateCb
+}
+
+// anchorTxNotification is a struct that holds all proof watch subscriptions for
+// a single anchor transaction. It is possible to have different callbacks for
+// different proofs within the same anchor transaction.
+type anchorTxNotification struct {
+	proofsRegistrations []*proofRegistration
+
+	cancel context.CancelFunc
+}
+
+// numProofs returns the total number of proofs that are being watched for this
+// anchor transaction.
+func (a *anchorTxNotification) numProofs() int {
+	return fn.Reduce(
+		a.proofsRegistrations,
+		func(agg int, reg *proofRegistration) int {
+			return agg + len(reg.proofs)
+		},
+	)
+}
+
+// firstProof returns the first proof that is being watched for this anchor
+// transaction.
+func (a *anchorTxNotification) firstProof() *proof.Proof {
+	for _, reg := range a.proofsRegistrations {
```

Comment:
Why can't this just be `a.proofsRegistrations[0]`? Or rather even just check the length first.